Bundle a .zip file; mount it with `@yarnpkg/fslib`

cspotcode opened the pull request:
**What this PR does / why we need it**:

Uses `@yarnpkg/fslib` both to create the bundle as a normal .zip and to mount it at runtime.

Still WIP; there's a bug where `fs.stat('/snapshot')` is failing even though reading contents of `/snapshot/*` is working.

**Which issue(s) this PR fixes**:

Fixes #740 

**Special notes for your reviewer**:

I want to share my work-in-progress and would appreciate hearing feedback.  Some big changes, and might be dropping support for older node versions.  So it's possible this will need to live as a fork of `nexe`.

Since the `patch.ts` is now a much larger bundle, I'm hoping I can bake it into the node binary and include it in the v8 snapshot.

calebboyd commented:
I'm thinking the next major is a rewrite anyways -- so this fits the bill to break the way its worked in the past!

calebboyd commented on src/fs/patch.ts:
```diff
@@ -1,398 +1,96 @@
+
+  const blob = Buffer.allocUnsafe(binary.layout.resourceSize)
+  const blobFd = realFs.openSync(binary.blobPath, 'r')
+  const bytesRead = realFs.readSync(
```
I'm concerned that there is too much being allocated here. Currently, the blobs remain on disk and aren't loaded into memory until read.. Do you know if there is anyway to base this off of a file/offset so we can keep the snapshot on disk?

calebboyd replied:
re: looks like ZipFS would need to support accepting an offset, which it does not currently.

cspotcode replied:
IIRC at the time I was reasoning that for common cases it needed to be pulled into memory anyway, so if we bite the bullet and do that up front, it'll be faster.

Do zips require an offset?  They're specifically designed to keep the table of contents at the end of the file, and I think all contents are referenced by relative offset from the table of contents.  I think a lot of self-extracting installers are a pre-built exe appended with a zip.

I did a quick test on my Linux box:
```
wget zip-file-from-somewhere.zip
cp `which bash` ./exe
cat zip-file-from-somewhere.zip >> ./exe
./exe -c 'echo foo' # works like bash
unzip -vl ./exe # also appears to be a valid zip file
```

calebboyd replied:
You're right that that is generally the common case. I don't have very much evidence but am hesitant to lose the advantage of direct file reads via the offset currently used.

cspotcode replied:
I think you may not need an offset to do direct file reads, due to the way zip files are structured.

calebboyd commented:
@cspotcode Did you have any luck going down the objcopy route?

I'm going to be paying down some of these issues in the near future and am trying to figure out the best approach here.

fslib is definitely the direction to go - But how to include the fakefs payload  _quickly_ in a pre-built binary, cross-platform is still stumping me a bit.
I'd also say that nexe has the requirement that the resulting binary/app be signable.

cspotcode commented:
It's been a while so I don't remember what I did with `objcopy`.  Looking back, I think the only reason I wanted to use `objcopy` is so that the zip payload was memory-mapped when the executable loaded, just like a raw data segment that had been compiled / linked into the executable at build time.  *Theoretically* this would be faster because the OS would memory-map the data along with the rest of the node executable, and then we could access it as RAM, create a `Buffer` instance pointing to the read-only data, etc.  I say "theoretically" because I don't understand the details well enough to know it would be faster.

If you don't care about that, then I think you can do `cat payload.zip >> executable` and it'll be a valid zip file and still executable.

calebboyd commented:
Yeah so the cat approach is effectively what is occurring now - but it has the caveat of being unsignable. 
So including the arbitrary payload in the binary as an additional section would be ideal - updating the memory map.

One approach I'm going to look into is is https://github.com/lief-project/LIEF but ideally I'd like to stick to non-native dependencies...

cspotcode commented:
You might be able to compile LIEF to WASM.

How does executable signing work for various executable formats?  What needs to change to sign the executable after it's been modified?  Googling around, [this page](https://blog.barthe.ph/2009/02/22/change-signed-executable/) suggests you can append arbitrary data to the digital certificate section and you only need to update that section's length, not recompute the cert.  But I haven't tested.

calebboyd commented:
>You might be able to compile LIEF to WASM.

That might be the best solution. I've bounced around various methods - and its easiest and best documented for PE executables for windows. But it is also relatively simple to modify the RC section pre-signature. I haven't had any luck discovering a solution for other binary formats (mach o and elf)

cspotcode commented:
Can we split this change into 2x tasks:
a) using zip payloads with current appending behavior
b) change how payload is added to support exe signing

If we split like this, then we can safely implement a) as long as we ensure that yarnpkg/libzip will support *eventually* implementing b).

This means we'll need to send a PR to yarnpkg/libzip which adds the ability to parse zips at an offset within a file.  I think this requires exposing [`zip_source_file_create`](https://libzip.org/documentation/zip_source_file_create.html) from libzip and adding offset fields to [`ZipPathOptions`](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-fslib/sources/ZipFS.ts#L28-L31)

calebboyd commented:
Yeah thats a great approach. The faster we can get rid of the hacked fs the better 😅

cspotcode commented:
Ok, cool.  I informally asked the yarn team if they would accept this PR, since it's adding a feature that they technically won't use themselves.  They said it's fine particularly if we include tests.

calebboyd commented:
Nice, I've only looked at it enough to see that it wasn't there when you initially did this work.

cspotcode commented:
I remember more of the `objcopy` stuff from my experiments.  The idea is: compile-in a pointer to a binary section, inject payload into the section, then use the pointer at startup to create an immutable `Buffer` and pass it to our bootstrapper.

C++ compilers let you declare exe sections via C++ and get pointers to the section as global variables.

The section can be expanded by `objcopy` to store both the bootstrapper `.js` and `.zip` payload, plus a simple header with offsets / sizes of each.

The pre-built `nexe` binary will have a small C++ modification.  It'll use the section pointer to create an immutable `Buffer` of the section and expose it to `_third_party_main.js`, which can invoke the bootstrapper `.js`.

Bootstrapper can be improved without recompiling the nexe binary, since it is bundled alongside the payload.  Also, it doesn't use extra fs calls at startup, since it takes advantage of whatever the OS does to memory-map the exe into memory.

Even if modifying the section requires re-running the linker, it's still just the linker.  So most compiler state can be prepared ahead of time, baked into a docker image, for example.

Some docs about creating sections via C++: https://stackoverflow.com/questions/3808053/how-to-get-a-pointer-to-a-binary-section-in-msvc
This may require our special section to appear last in the binary so that expanding it without re-running the linker does not break the pointers.

calebboyd commented:
Yeah, I that would work. It might even be worth it to take a hex editor to each bin before/after manually performing the operations for each platform -- for the sole purpose of keeping the actual tool lightweight. I'd like to keep everything as lightweight as can be - without needing a linker or requiring docker to use it. 

There are [fun advantages](https://calebboyd.github.io/xbin/) to the js ecosystem 

One approach might also be (not sure if possible -) but a legitimate "onionduke" modification... I'm not sure if that even applies pre-signed binary 🤔

src: https://github.com/secretsquirrel/the-backdoor-factory

cspotcode commented:
>  It might even be worth it to take a hex editor to each bin before/after manually performing the operations for each platform

Ooh yeah, great idea.  I bet we could automate the process.  So CI uses objcopy or something similar to perform the operations with a test payload, then a script analyzes the differences before/after and emits the necessary information alongside the prebuilt binary.  Then nexe only needs the prebuilt binary and the analysis information to embed payloads.

cspotcode commented:
Is this an accurate description of the middleware flow in nexe?  Ordered temporally from top to bottom, using indentation to indicate which middleware is executing a particular task.

```
- clean
- resource
- cli
  - next()
      - bundle
      - shim
          - next()
              - download
              - artifacts
                  - write temp files to node source ??
                  - next()
                      - ...built-in patches
                      - ...options.patches: user-provided patches
                      - ...plugins
                  - patched files written to node source ??
          - shims.push() 3x shims
  - write deliverable to disk


```

calebboyd commented:
Yeah, for artifacts - the temp directory just holds clean files. so any patches run against the same input on sequential builds can use them.

Since moving the bundling to resolve-dependencies the middleware layout for this project makes less sense...

Julusian commented:
Whats the current status of this?
I would really like to switch a project to nexe from electron, but am blocked on the lack of worker_threads which I believe this will lay the groundwork for

cspotcode commented:
If you would like to continue the work on this pull request, feel free.  I
likely will not have time for it in the near future.

On Sun, Mar 14, 2021, 8:40 AM Julian Waller ***@***.***>
wrote:

> Whats the current status of this?
> I would really like to switch a project to nexe from electron, but am
> blocked on the lack of worker_threads which I believe this will lay the
> groundwork for
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/nexe/nexe/pull/741#issuecomment-798900143>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC35OG33E33ECJHF5CXWE3TDSVEXANCNFSM4LO25L5A>
> .
>

cspotcode commented:
Cross-referencing this: https://github.com/cspotcode/nexe/pull/1
Someone else sent me some improvements to this PR which I haven't had a chance to read.

calebboyd commented:
Sorry I let this get stale! Life's been busy! Moving forward with these as picked up by @bruce-one
